app/models/rbac.rb

```diff
@@ -109,28 +109,31 @@ def self.rbac_instance(obj)
     nil
   end
 
-  def self.get_self_service_objects(user_or_group, klass)
-    return nil if user_or_group.nil? || !user_or_group.self_service? || !(klass < OwnershipMixin)
+  def self.get_self_service_objects(user, miq_group, klass)
+    return nil if miq_group.nil? || !miq_group.self_service? || !(klass < OwnershipMixin)
+
+    # for limited_self_service, use user's resources, not user.current_group's resources
+    # for reports (user = nil), still use miq_group
+    miq_group = nil if user && miq_group.limited_self_service?
 
     # Get the list of objects that are owned by the user or their LDAP group
-    cond = user_or_group.limited_self_service? ? klass.conditions_for_owned(user_or_group) : klass.conditions_for_owned_or_group_owned(user_or_group)
-    klass.select(minimum_columns_for(klass)).where(cond)
+    klass.user_or_group_owned(user, miq_group).select(minimum_columns_for(klass))
   end
 
   # @return nil if no objects are owned by self service or user not a selfservice user
-  # @return [Array<Integer>] object_ids owned by a user or group
-  def self.get_self_service_object_ids(user_or_group, klass)
-    targets = get_self_service_objects(user_or_group, klass)
-    targets = targets.reorder(nil).collect(&:id) if targets.respond_to?(:collect)
-    targets
+  # @return [Array<Integer>, nil] object_ids owned by a user or group
+  def self.get_self_service_object_ids(user, miq_group, klass)
+    targets = get_self_service_objects(user, miq_group, klass)
+    targets.reorder(nil).pluck(:id) if targets
   end
 
-  def self.calc_filtered_ids(scope, user_filters, user_or_group = nil)
+  def self.calc_filtered_ids(scope, user_filters, user, miq_group)
     klass = scope.respond_to?(:klass) ? scope.klass : scope
-    u_filtered_ids = get_self_service_object_ids(user_or_group, klass)
+    u_filtered_ids = get_self_service_object_ids(user, miq_group, klass)
     b_filtered_ids = get_belongsto_filter_object_ids(klass, user_filters['belongsto'])
     m_filtered_ids = get_managed_filter_object_ids(klass, scope, user_filters['managed'])
-    d_filtered_ids = user_filters['ids_via_descendants']
+    d_filtered_ids = ids_via_descendants(rbac_class(klass), user_filters['match_via_descendants'],
```

Comment:
@gtanzillo The intent was to move the descendant logic lower down the tree, closer to the id lookup clauses. The next step is to not bring back the ids at all but rather perform the 4 queries in the database.

```diff
@Rbac.search
-    user_filters['ids_via_descendants'] = ids_via_descendants(rbac_class(klass), options.delete(:match_via_descendants), :user => user, :miq_group => miq_group)
+    user_filters['match_via_descendants'] = options.delete(:match_via_descendants)
@Rbac.calc_filtered_ids
-    d_filtered_ids = user_filters['ids_via_descendants']
+    d_filtered_ids = ids_via_descendants(rbac_class(klass), user_filters['match_via_descendants'],
+                                         :user => user, :miq_group => miq_group)
```

This was performed in [Rbac pass match_via_descendants not ids](https://github.com/ManageIQ/manageiq/pull/7141/commits/52dc165d79522baf943c02c32b302911ac9fec96)